aiida/repository/backend/disk_object_store.py

```diff
@@ -118,3 +118,63 @@ def get_object_hash(self, key: str) -> str:
         if self.container.hash_type != 'sha256':
             return super().get_object_hash(key)
         return key
+
+    def maintain(self, full: bool = False, **kwargs) -> dict:
+        pack_loose = True
+        do_repack = full
+        clean_storage = full
+        do_vacuum = full
+
+        operation_results = {}
+
+        if 'override_pack_loose' in kwargs:
+            pack_loose = kwargs.get('override_pack_loose')  # type: ignore
+
+        if 'override_do_repack' in kwargs:
+            do_repack = kwargs.get('override_do_repack')  # type: ignore
+
+        if 'override_clean_storage' in kwargs:
+            clean_storage = kwargs.get('override_clean_storage')  # type: ignore
+
+        if 'override_do_vacuum' in kwargs:
+            do_vacuum = kwargs.get('override_do_vacuum')  # type: ignore
```

Comment:
I had to disable mypy here because I am getting errors I don't understand:

```
aiida/repository/backend/disk_object_store.py:131: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "bool") [assignment] 
aiida/repository/backend/disk_object_store.py:134: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "bool") [assignment] 
aiida/repository/backend/disk_object_store.py:137: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "bool") [assignment] 
aiida/repository/backend/disk_object_store.py:140: error: Incompatible types in assignment (expression has type "Optional[Any]", variable has type "bool") [assignment] 
```

Reply:
Problem here is that those variables (`pack_loose` etc.) are declared as having type `bool`. But the `get` method of a dictionary will return `Optional[Any]`, i.e. any type or `None`. This means you can assign `None` or any other value to what is supposed to be a bool.

Why do you use `**kwargs` in the method signature? If you just explicitly define them, you don't have this problem. Plus, you can get rid of all these initialization rules and just define proper defaults. In any case, I would have used the following syntax:
```python
pack_loose = kwargs.get('override_pack_loose', True)
```
But really the question is why you chose this design anyway? Why have the caller use the argument prefixed with `override_` if you are going to change it internally. Doesn't it make way more sense to just have `pack_loose` as the method argument?

Reply:
> Problem here is that those variables (`pack_loose` etc.) are declared as having type `bool`.

I mean, where? Those are internal variables, I never explicitly tell mypy they are anything. Does it interpret this from me doing `pack_loose = True` or `do_repack = full`? Does that mean I can no longer re-assign internal variables to "different" types?

> Plus, you can get rid of all these initialization rules and just define proper defaults. In any case, I would have used the following syntax:
> ```python
> pack_loose = kwargs.get('override_pack_loose', True)
> ```

Mmm this is true. This was a leftover of a previous way of setting the initial values, which I defaulted to the bools for a fast maintenance and then changed them if the user was doing `full`. But now that I use the `full` flag to assign it as the default value itself I guess I can use that with the overrides as well.

> Why do you use **kwargs in the method signature? If you just explicitly define them, you don't have this problem.
> ...
> Why have the caller use the argument prefixed with override_ if you are going to change it internally. Doesn't it make way more sense to just have pack_loose as the method argument?

I think the answer to both of these is the same: the method first needs to decide if it will set these values for a `full` or a `partial` maintenance, and afterwards the overrides can change any of those values independently.

If I was setting defaults in the signature then I would have to choose one (`full` or a `partial`) and then in the body detect if the other is being performed and adapt the values. In that case, I would then be overriding the more specific "overrides" requested by the user. Although you could say that technically the same resulting procedures are available either case, I think it makes more sense that the more specific indications gets respected instead of it being stomp over the "general" setup.

Reply:
> I mean, where? Those are internal variables, I never explicitly tell mypy they are anything. Does it interpret this from me doing pack_loose = True or do_repack = full?

Yes, by assigning a bool to those variables, `mypy` will interpret that this variable is supposed to represents a boolean. So when you then assign `None` to it, it will raise a warning. Note that it _is_ just a warning and you may want to allow this behavior, in which case ignoring the warning is reasonable. However, note that in this case, it probably _is_ a bug. You don't type check and so you could assign anything to those variables, for example `None`, and this could have unintended consequences downstream.

> I think the answer to both of these is the same: the method first needs to decide if it will set these values for a full or a partial maintenance, and afterwards the overrides can change any of those values independently.

Fair enough, but I think this really is just a problem of the design of the signature to begin with. I don't think the concept of `full` and `partial` should be baked into maintain. That seems to be a high-level concept that the CLI will have, but the lower-level interface doesn't need this. I would simplify this to the following signature:
```python
def maintain(self, pack_loose = True, repack = False, clean_storage = False, vacuum = False) -> dict:
```
i.e. with defaults that correspond to your `partial` and then the CLI can call this with all `True` if the user requests a `full` maintenance.

Reply:
> Yes, by assigning a bool to those variables, `mypy` will interpret that this variable is supposed to represents a boolean. So when you then assign `None` to it, it will raise a warning. Note that it _is_ just a warning and you may want to allow this behavior, in which case ignoring the warning is reasonable. However, note that in this case, it probably _is_ a bug. You don't type check and so you could assign anything to those variables, for example `None`, and this could have unintended consequences downstream.

I think I get the distinction you are trying to make with it being a superfluous warning or an actual bug catcher. But these are all runtime user end inputs, so the typing is not enforced and adding it wouldn't really catch anything. Would mypy recognize if I do some manual typechecking with `isinstance` to guard for this or what is the option?


> Fair enough, but I think this really is just a problem of the design of the signature to begin with. I don't think the concept of `full` and `partial` should be baked into maintain. That seems to be a high-level concept that the CLI will have, but the lower-level interface doesn't need this. I would simplify this to the following signature:
> ```python
> def maintain(self, pack_loose = True, repack = False, clean_storage = False, vacuum = False) -> dict:
> ```
> i.e. with defaults that correspond to your `partial` and then the CLI can call this with all `True` if the user requests a `full` maintenance.

Normally I agree, but here the idea to call `repository_maintain` with `full=True` or `full=False` is that you don't need to know the specifics of the backend, only the implemented backend knows what `full` means to them. The `repository_maintain` doesn't know what the underlying backend is and thus what its specific switches are: the `full` keyword is the only option we require the repository backend to implement, the rest are optional switches that would be passed from the user to the backend transparently (i.e. `repository_maintain` just passes them, doesn't know what they do), with the understanding that is the user the one who know what specific backend is using and thus the keywords that are available.

P.S. I have not yet applied the modification suggested above in case you are in the middle of reviewing so that it doesn't change stuff. If you haven't started I add it now and push it.

Reply:
I'll mark this as resolved just because I modified this as you suggested in the first comment and now pylint doesn't complain anymore (thanks!). If you still disagree about the general design feel free to keep replying though.